source/agora/consensus/Ledger.d

```diff
@@ -641,11 +646,13 @@ public class Ledger
                         ValidatorInfo[] validators = this.getValidators(header.height)
                             .enumerate.filter!(en => header.validators[en.index]).map!(en => en.value).array;
 
-                        auto fees = this.fee_man.getBlockFees(header.height);
-                        auto data_fees = this.fee_man.getBlockDataFees(header.height);
+                        // Calculate the block rewards using the percentage of validators who signed
+                        auto rewards = this.rewards.calculateBlockRewards(Height(height - this.params.PayoutPeriod),
+                            header.validators.percentage());
+
+                        // Divide up the validator fees and rewards based on stakes
+                        Amount[] val_payouts = this.fee_man.getValidatorPayouts(header.height, rewards, validators);
```

Comment:
As a side note, all those functions taking arrays as parameter lead to extra allocation, which should be done by ranges instead.

Reply:
Good point, I will address in new PR.